File: GUI.java
package carcassonne;

import java.awt.*;
import java.awt.geom.*;
import java.awt.image.*;
import javax.swing.*;
import javax.swing.table.*;
import java.lang.Integer;

public class GUI {
    static Tile[][] strdata;
    static Tile[][] data;
    static JScrollPane sp;
    // Constructor
    public static void gui(Board board, JFrame f, int x, int y) {
        // Table
        JTable j;

        // Frame Title
        f.getContentPane().setLayout(new BorderLayout());
        f.setTitle("Carcassonne");

        // Data to be displayed in the JTable
        data = board.board;
        strdata = new Tile[133][133];
        for (int i = 0; i < 133; i++) {
            for (int k = 0; k < 133; k++) {
                strdata[k][i] = data[i][k];
            }
        }


        // Column Names
        String[] columnNames = new String[133];
        columnNames[0] = "Row #";
        for (int i = 0; i < 133; i++) {
            columnNames[i] = Integer.toString(i);
        }

        // Initializing the JTable
        j = new JTable(strdata, columnNames);
        j.setBounds(133, 133, 500, 133);
        j.setRowHeight(222);
        j.setAutoResizeMode(JTable.AUTO_RESIZE_OFF);
        for (int i = 0; i < 133; i++) {
            j.getColumnModel().getColumn(i).setCellRenderer(new ImageRenderer());
            j.getColumnModel().getColumn(i).setPreferredWidth(222);
        }

        // adding it to JScrollPane
        sp = new JScrollPane(j);
        
        //f.add(sp);
        f.getContentPane().add(sp);
        // Frame Size
        
        // Frame Visible = true
        f.setVisible(true);

        sp.getVerticalScrollBar().setValue(y);
        sp.getHorizontalScrollBar().setValue(x);
    }
}

class ImageRenderer extends DefaultTableCellRenderer {
    JLabel lbl = new JLabel();

    ImageIcon blank = new ImageIcon(getClass().getResource("res/emptytile.png"));
    ImageIcon plusroad = new ImageIcon(getClass().getResource("res/+road.png"));
    ImageIcon oneedgecity = new ImageIcon(getClass().getResource("res/1edgecity.png"));
    ImageIcon onenot = new ImageIcon(getClass().getResource("res/1not.png"));
    ImageIcon oneroad = new ImageIcon(getClass().getResource("res/1road.png"));
    ImageIcon twoadjcity = new ImageIcon(getClass().getResource("res/2adjacentcity.png"));
    ImageIcon allcity = new ImageIcon(getClass().getResource("res/allcity.png"));
    ImageIcon bowtiecity = new ImageIcon(getClass().getResource("res/bowtiecity.png"));
    ImageIcon bowtiefield = new ImageIcon(getClass().getResource("res/bowtiefield.png"));
    ImageIcon cornercity = new ImageIcon(getClass().getResource("res/cornercity.png"));
    ImageIcon cornercityroad = new ImageIcon(getClass().getResource("res/cornercityroad.png"));
    ImageIcon edgecityroadleft = new ImageIcon(getClass().getResource("res/edgecityroadleft.png"));
    ImageIcon edgecityroadright = new ImageIcon(getClass().getResource("res/edgecityroadright.png"));
    ImageIcon edgecityroadstraight = new ImageIcon(getClass().getResource("res/edgecityroadstraight.png"));
    ImageIcon edgecityroadt = new ImageIcon(getClass().getResource("res/edgecityroadt.png"));
    ImageIcon lroad = new ImageIcon(getClass().getResource("res/Lroad.png"));
    ImageIcon straightroad = new ImageIcon(getClass().getResource("res/straightroad.png"));
    ImageIcon troad = new ImageIcon(getClass().getResource("res/Troad.png"));
    ImageIcon meeple = new ImageIcon(getClass().getResource("res/meeple.png"));

    public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column) {
        // if (value.equals("emptytile.png"))
        //     lbl.setIcon(blank);
        Icon img = blank;
        if (((Tile) value).type.equals("")) {
            lbl.setIcon(blank);
            lbl.setText("(" + Integer.toString(column) + ", " + Integer.toString(row) + ")");
            return lbl;
        }
        if (((Tile) value).type.equals("edgecityroadstraight")) {
            img = new ImageIcon(edgecityroadstraight.getImage());
        }
        else if (((Tile) value).type.equals("allcity")) {
            img = new ImageIcon(allcity.getImage());
        }
        else if (((Tile) value).type.equals("1not")) {
            img = new ImageIcon(onenot.getImage());
        }
        else if (((Tile) value).type.equals("1road")) {
            img = new ImageIcon(oneroad.getImage());
        }
        else if (((Tile) value).type.equals("cornercity")) {
            img = new ImageIcon(cornercity.getImage());
        }
        else if (((Tile) value).type.equals("cornercityroad")) {
            img = new ImageIcon(cornercityroad.getImage());
        }
        else if (((Tile) value).type.equals("bowtiecity")) {
            img = new ImageIcon(bowtiecity.getImage());
        }
        else if (((Tile) value).type.equals("2adjacentcity")) {
            img = new ImageIcon(twoadjcity.getImage());
        }
        else if (((Tile) value).type.equals("bowtiefield")) {
            img = new ImageIcon(bowtiefield.getImage());
        }
        else if (((Tile) value).type.equals("1edgecity")) {
            img = new ImageIcon(oneedgecity.getImage());
        }
        else if (((Tile) value).type.equals("edgecityroadleft")) {
            img = new ImageIcon(edgecityroadleft.getImage());
        }
        else if (((Tile) value).type.equals("edgecityroadright")) {
            img = new ImageIcon(edgecityroadright.getImage());
        }
        else if (((Tile) value).type.equals("edgecityroadt")) {
            img = new ImageIcon(edgecityroadt.getImage());
        }
        else if (((Tile) value).type.equals("straightroad")) {
            img = new ImageIcon(straightroad.getImage());
        }
        else if (((Tile) value).type.equals("Lroad")) {
            img = new ImageIcon(lroad.getImage());
        }
        else if (((Tile) value).type.equals("Troad")) {
            img = new ImageIcon(troad.getImage());
        }
        else if (((Tile) value).type.equals("+road")) {
            img = new ImageIcon(plusroad.getImage());
        }
        lbl.setText("(" + Integer.toString(column) + ", " + Integer.toString(row) + ")");
        // img = new RotatedIcon(img, Rotate.ABOUT_CENTER);
        // ((RotatedIcon) img).setDegrees(((Tile) value).rotations * 90.0);
        BufferedImage combined = new BufferedImage(220, 220, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g = combined.createGraphics();
        AffineTransform old = g.getTransform();
        g.rotate(((Tile) value).rotations * Math.PI / 2, 110, 110);
        g.drawImage(((ImageIcon) img).getImage(), 0, 0, null);
        g.setTransform(old);
        if (((Tile) value).meeple[14] != -1) {
            int[] pos = meeplePos(((Tile) value).meeple[14]);
            g.drawImage(meeple.getImage(), pos[0], pos[1], null);
        }
        g.dispose();
        lbl.setIcon(new ImageIcon(combined));
        return lbl;
    }

    int[] meeplePos(int pos) {
        int[] xy = new int[2];
        switch(pos) {
            case 0:
                xy = new int[] {42, 14};
                break;
            case 1:
                xy = new int[] {110, 14};
                break;
            case 2:
                xy = new int[] {178, 14};
                break;
            case 3:
                xy = new int[] {206, 42};
                break;
            case 4:
                xy = new int[] {206, 110};
                break;
            case 5:
                xy = new int[] {206, 178};
                break;
            case 6:
                xy = new int[] {178, 206};
                break;
            case 7:
                xy = new int[] {110, 206};
                break;
            case 8:
                xy = new int[] {42, 206};
                break;
            case 9:
                xy = new int[] {14, 178};
                break;
            case 10:
                xy = new int[] {14, 110};
                break;
            case 11:
                xy = new int[] {14, 42};
                break;
            default:
                xy = new int[] {0, 0};
                break;
        }
        xy[0] -= 14;
        xy[1] -= 14;
        return xy;
    }
}

File: GUI.js

class GUI {
    static strdata;
    static data;
    static sp;
  
    static gui(board) {
      let j;
  
      this.data = board.board;
      this.strdata = new Array(133).fill(0).map(() => new Array(133));
      for (let i = 0; i < 133; i++) {
        for (let k = 0; k < 133; k++) {
          this.strdata[k][i] = this.data[i][k];
        }
      }
  
      const columnNames = new Array(133);
      columnNames[0] = "Row #";
      for (let i = 0; i < 133; i++) {
        columnNames[i] = i.toString();
      }
  
      j = new JTable(this.strdata, columnNames);
      j.setBounds(133, 133, 500, 133);
      j.setRowHeight(222);
      j.setAutoResizeMode(JTable.AUTO_RESIZE_OFF);
      for (let i = 0; i < 133; i++) {
        j.getColumnModel().getColumn(i).setCellRenderer(new ImageRenderer());
        j.getColumnModel().getColumn(i).setPreferredWidth(222);
      }
  
      this.sp = new JScrollPane(j);
  
      return this.sp;
    }
  }
  
  class ImageRenderer extends DefaultTableCellRenderer {
    lbl = new JLabel();
  
    blank = new ImageIcon(getClass().getResource("res/emptytile.png"));
    plusroad = new ImageIcon(getClass().getResource("res/+road.png"));
    oneedgecity = new ImageIcon(getClass().getResource("res/1edgecity.png"));
    onenot = new ImageIcon(getClass().getResource("res/1not.png"));
    oneroad = new ImageIcon(getClass().getResource("res/1road.png"));
    twoadjcity = new ImageIcon(getClass().getResource("res/2adjacentcity.png"));
    allcity = new ImageIcon(getClass().getResource("res/allcity.png"));
    bowtiecity = new ImageIcon(getClass().getResource("res/bowtiecity.png"));
    bowtiefield = new ImageIcon(getClass().getResource("res/bowtiefield.png"));
    cornercity = new ImageIcon(getClass().getResource("res/cornercity.png"));
    cornercityroad = new ImageIcon(getClass().getResource("res/cornercityroad.png"));
    edgecityroadleft = new ImageIcon(getClass().getResource("res/edgecityroadleft.png"));
    edgecityroadright = new ImageIcon(getClass().getResource("res/edgecityroadright.png"));
    edgecityroadstraight = new ImageIcon(getClass().getResource("res/edgecityroadstraight.png"));
    edgecityroadt = new ImageIcon(getClass().getResource("res/edgecityroadt.png"));
    lroad = new ImageIcon(getClass().getResource("res/Lroad.png"));
    straightroad = new ImageIcon(getClass().getResource("res/straightroad.png"));
    troad = new ImageIcon(getClass().getResource("res/Troad.png"));
    meeple = new ImageIcon(getClass().getResource("res/meeple.png"));
    meepley = new ImageIcon(getClass().getResource("res/meepleyellow.png"));
    meepleg = new ImageIcon(getClass().getResource("res/meeplegreen.png"));
  
    getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column) {
      let img = this.blank;
      if (value.type === "") {
        this.lbl.setIcon(this.blank);
        this.lbl.setText(`(${column}, ${row})`);
        return this.lbl;
      }
      if (value.type === "edgecityroadstraight") {
        img = new ImageIcon(this.edgecityroadstraight.getImage());
      } else if (value.type === "allcity") {
        img = new ImageIcon(this.allcity.getImage());
      } else if (value.type === "1not") {
        img = new ImageIcon(this.onenot.getImage());
      } else if (value.type === "1road") {
        img = new ImageIcon(this.oneroad.getImage());
      } else if (value.type === "cornercity") {
        img = new ImageIcon(this.cornercity.getImage());
      } else if (value.type === "cornercityroad") {
        img = new ImageIcon(this.cornercityroad.getImage());
      } else if (value.type === "bowtiecity") {
        img = new ImageIcon(this.bowtiecity.getImage());
      } else if (value.type === "2adjacentcity") {
        img = new ImageIcon(this.twoadjcity.getImage());
      } else if (value.type === "bowtiefield") {
        img = new ImageIcon(this.bowtiefield.getImage());
      } else if (value.type === "1edgecity") {
        img = new ImageIcon(this.oneedgecity.getImage());
      } else if (value.type === "edgecityroadleft") {
        img = new ImageIcon(this.edgecityroadleft.getImage());
      } else if (value.type === "edgecityroadright") {
        img = new ImageIcon(this.edgecityroadright.getImage());
      } else if (value.type === "edgecityroadt") {
        img = new ImageIcon(this.edgecityroadt.getImage());
      } else if (value.type === "straightroad") {
        img = new ImageIcon(this.straightroad.getImage());
      } else if (value.type === "Lroad") {
        img = new ImageIcon(this.lroad.getImage());
      } else if (value.type === "Troad") {
        img = new ImageIcon(this.troad.getImage());
      } else if (value.type === "+road") {
        img = new ImageIcon(this.plusroad.getImage());
      }
      this.lbl.setText(`(${column}, ${row})`);
      const combined = new BufferedImage(220, 220, BufferedImage.TYPE_INT_ARGB);
      const g = combined.createGraphics();
      const old = g.getTransform();
      g.rotate(value.rotations * Math.PI / 2, 110, 110);
      g.drawImage(img.getImage(), 0, 0, null);
      g.setTransform(old);
      if (value.meeple[14] !== -1) {
        const pos = this.meeplePos(value.meeple[14]);
        if (value.meeple[0] === 0) {
          g.drawImage(this.meeple.getImage(), pos[0], pos[1], null);
        } else if (value.meeple[0] === 1) {
          g.drawImage(this.meepley.getImage(), pos[0], pos[1], null);
        } else if (value.meeple[0] === 2) {
          g.drawImage(this.meepleg.getImage(), pos[0], pos[1], null);
        }
      }
      g.dispose();
      this.lbl.setIcon(new ImageIcon(combined));
      return this.lbl;
    }
  
    meeplePos(pos) {
      switch (pos) {
        case 0:
          return [42, 14];
        case 1:
          return [110, 14];
        case 2:
          return [178, 14];
        case 3:
          return [206, 42];
        case 4:
          return [206, 110];
        case 5:
          return [206, 178];
        case 6:
          return [178, 206];
        case 7:
          return [110, 206];
        case 8:
          return [42, 206];
        case 9:
          return [14, 178];
        case 10:
          return [14, 110];
        case 11:
          return [14, 42];
        default:
          return [0, 0];
      }
    }
  }
  
  